File: proyectoRuby/lib/tablero.rb
# encoding :UTF-8
require_relative 'casilla'

module Civitas
class Tablero
  
  # Constructor
  def initialize(n)
    if n > 1
      @numCasillaCarcel = n
    else
      @numCasillaCarcel = 1
    end
      
    @casillas = []
    @porSalida = 0
    @tieneJuez = false
    
    # Añade la casilla de salida
    salida = Casilla.new("Salida")
    @casillas.push(salida)
  end
  
  # Métodos
  private
  def correcto(n=0)
    tablero_ok = (@numCasillaCarcel < @casillas.length) && @tieneJuez
    tablero_ok && (n < @casillas.length) && (n >= 0)
  end
  
  public
  def añade_casilla(c)
    num_casillas = @casillas.length
    if (num_casillas == @numCasillaCarcel)
      @casillas.push(Casilla.new("Cárcel"))
    end
    
    @casillas.push(c)
    if (num_casillas == @numCasillaCarcel)
      @casillas.push(Casilla.new("Cárcel"))
    end
  end
  
  def añade_juez
    if !@tieneJuez
      @casillas.push(Casilla.new("Juez"))
      @tieneJuez = true
    end
  end
  
  def nueva_posicion(actual, tirada)
    if !correcto(actual)
      -1
    else
      nueva_posi = (actual + tirada) % @casillas.length
      
      if correcto(nueva_posi)
        if nueva_posi != (actual + tirada)
          @porSalida += 1
        end
        
        nueva_posi
      end
    end
  end
  
  def calcular_tirada(origen, destino)
    resta = destino - origen
    
    if resta >= 0
      resta
    else
      resta + @casillas.length
    end
  end
  
  # Consultores
  attr_reader :numCasillaCarcel
  
  def get_por_salida
    if @porSalida > 0
      temp = @porSalida
      @porSalida -= 1
      temp
    else
      @porSalida
    end
  end
  
  def get_casilla(n)
    if correcto(n)
      @casillas[n]
    else
      nil
    end
  end  
end
end

File: CivitasJava/civitas/Tablero.java
package civitas;

import java.util.ArrayList;

public class Tablero{
    // Atributos
    private int numCasillaCarcel;
    ArrayList<Casilla> casillas;
    int porSalida;
    boolean tieneJuez;
    
    // Constructor
    public Tablero( int i ){
        if (i >= 1)
            numCasillaCarcel = i;
        else
            numCasillaCarcel = 1;
        
        casillas = new ArrayList<>();
        Casilla casilla_salida = new Casilla("Salida");
        casillas.add(0, casilla_salida);
        
        porSalida = 0;
        tieneJuez = false;
    }
    
    // Métodos
    private boolean correcto(){
        if ( casillas.size() > numCasillaCarcel && tieneJuez )
            return true;
        else
            return false;
    }
    
    private boolean correcto(int numCasilla){
        if ( correcto() && numCasilla < casillas.size() )
            return true;
        else
            return false;
    }
    
    /**
     * Consultor para el atributo numCasillaCarcel
     * @return número entero donde se ubica la cárcel en el tablero
     */
    int getCarcel(){
        return numCasillaCarcel;
    }
    
    int getPorSalida(){
        if (porSalida > 0){
            int temp = porSalida;
            porSalida--;
            return temp;
        }else{
            return porSalida;
        }
    }
    
    void añadeCasilla( Casilla casilla ){
        if ( casillas.size() == numCasillaCarcel ){
            Casilla temp = new Casilla("Cárcel");
            casillas.add(temp);
        };
        
        casillas.add(casilla);
        
        if ( casillas.size() == numCasillaCarcel ){
            Casilla temp = new Casilla("Cárcel");
            casillas.add(temp);
        };        
    }
    
    void añadeJuez( ){
        if (!tieneJuez){
            casillas.add(new CasillaJuez(numCasillaCarcel, "JUEZ"));
            tieneJuez = true;
        }
    }
    
    Casilla getCasilla (int numCasilla){
        if ( correcto(numCasilla) )
            return casillas.get(numCasilla);
        else
            return null;
    }
    
    int nuevaPosicion (int actual, int tirada){
        if ( !correcto() )
            return -1;
        else{
            int nuevaPosi = (actual + tirada) % casillas.size();
            if ( correcto(nuevaPosi) ){
                if (nuevaPosi != (actual + tirada) )
                    porSalida++;
                
                return nuevaPosi;
            }
        }
            
        return -1;
    }
    
    int calcularTirada (int origen, int destino){
        int resta = destino - origen;
        
        if ( resta >= 0)
            return resta;
        else
            return resta + casillas.size();
    }    
}
